### PR TITLE
fix(api,vcs,repositories): improve migrate ascode

### PR DESCRIPTION
**Pull request by richardlt**
1. Description
Remove usage of webworker to pull operation.
Improve errors (ex: missing auth to push).
Allow to reuse a existing branch.
Partially implement Gitlab PullRequest client.
Fix remove ascode if migration PR closed.

1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12831.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12831.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12833.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12833.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12845.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12845.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/api/ascode/sync/sync.go [hunk 1]
```diff
@@ -62,6 +62,21 @@ func SyncAsCodeEvent(ctx context.Context, db *gorp.DbMap, store cache.Store, pro
 			log.Debug("Pull request %s #%d not found", app.RepositoryFullname, int(ascodeEvt.PullRequestID))
 		}
 
+		// If the PR was deleted or closed we should remove the ascode data from the workflow, else if merged we want to set this value.
+		if ascodeEvt.Migrate && len(ascodeEvt.Data.Workflows) == 1 {
+			for id := range ascodeEvt.Data.Workflows {
+				if prNotFound || pr.Closed {
+					if err := workflow.UpdateFromRepository(tx, id, ""); err != nil {
```
Comment (sguiheux) [inline engine/api/ascode/sync/sync.go]: do not update is workflow is not merged

File: engine/repositories/processor_push.go [hunk 2]
```diff
@@ -24,29 +34,28 @@ func (s *Service) processPush(ctx context.Context, op *sdk.Operation) error {
...
 	}
 
-	// Switch to default branch
-	if currentBranch != op.RepositoryInfo.DefaultBranch {
-		if err := gitRepo.FetchRemoteBranch("origin", op.RepositoryInfo.DefaultBranch); err != nil {
-			log.Error(ctx, "Repositories> processPush> Checkout to default branch> [%s] error %v", op.UUID, err)
-			return sdk.WrapError(err, "Checkout to default branch> [%s] error %v", op.UUID, err)
+	// Switch to target branch
+	if currentBranch != op.Setup.Push.FromBranch {
+		if err := gitRepo.CheckoutNewBranch(op.Setup.Push.FromBranch); err != nil {
```
Comment (sguiheux) [inline engine/repositories/processor_push.go]: is it not better to go to default branch, git pull and then create the new branch?

File: engine/vcs/gitlab/client_pull_request.go [hunk 3]
```diff
@@ -23,5 +50,39 @@ func (c *gitlabClient) PullRequestComment(context.Context, string, sdk.VCSPullRe
 
 // PullRequestCreate create a new pullrequest
 func (c *gitlabClient) PullRequestCreate(ctx context.Context, repo string, pr sdk.VCSPullRequest) (sdk.VCSPullRequest, error) {
-	return sdk.VCSPullRequest{}, fmt.Errorf("not yet implemented")
+	mr, _, err := c.client.MergeRequests.CreateMergeRequest(repo, &gitlab.CreateMergeRequestOptions{
```
Comment (sguiheux) [inline engine/vcs/gitlab/client_pull_request.go]: it's the user that link the project to gitlab who create the pullrequest? is it possible to grant another user (like other vcs ) ?

File: ui/src/app/shared/modal/save-as-code/update.as.code.component.ts [hunk 4]
```diff
@@ -76,63 +82,39 @@ export class UpdateAsCodeComponent {
...
+
+    startPollingOperation() {
+        this.pollingOperationSub = Observable.interval(1000).subscribe(() => {
+            this._workflowService.getAsCodeOperation(this.project.key, this.name, this.ope.uuid)
+                .pipe(finalize(() => this._cd.markForCheck()))
+                .subscribe(o => {
+                    this.ope = o;
+                    if (this.ope.status > 1) {
+                        this.loading = false;
+                        this.pollingOperationSub.unsubscribe();
```
Comment (sguiheux) [inline ui/src/app/shared/modal/save-as-code/update.as.code.component.ts]: I Already have pb with calling unsubscribe in subscribe method.
I think, best way is doing something like that:  
.first(o => o.status > 1) .  and you will not need to take care of subscription

File: ui/src/assets/i18n/en.json [hunk 5]
```diff
@@ -125,6 +125,7 @@
   "btn_logs": "Display logs",
   "btn_rename": "Rename",
   "btn_resync_repo": "Resynchronize",
```
Comment (sguiheux) [inline ui/src/assets/i18n/en.json]: need delete line 127?

File: ui/src/assets/i18n/fr.json [hunk 6]
```diff
@@ -147,6 +147,7 @@
   "btn_next": "Suivant",
   "btn_rename": "Renommer",
   "btn_resync_repo": "Resynchroniser",
```
Comment (sguiheux) [inline ui/src/assets/i18n/fr.json]: need delete line 149?

Comment (richardlt) [reply ui/src/assets/i18n/fr.json]: no

File: ui/src/assets/i18n/en.json (hunk 5 above)
Comment (richardlt) [reply ui/src/assets/i18n/en.json]: no

File: engine/vcs/gitlab/client_pull_request.go (hunk 3 above)
Comment (richardlt) [reply engine/vcs/gitlab/client_pull_request.go]: Will do in another PR

File: engine/repositories/processor_push.go (hunk 2 above)
Comment (richardlt) [reply engine/repositories/processor_push.go]: I changed the way it works to remove usage of FetchRemoteBranch func from go-repo. Now there is a hars reset to origin/defaultBranch if the branch already exists locally but was never pushed.

File: engine/api/ascode/sync/sync.go (hunk 1 above)
Comment (richardlt) [reply engine/api/ascode/sync/sync.go]: done

File: ui/src/app/shared/modal/save-as-code/update.as.code.component.ts (hunk 4 above)
Comment (richardlt) [reply ui/src/app/shared/modal/save-as-code/update.as.code.component.ts]: Done using mergeMap and first

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12858.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12858.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12880.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#12880.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Pull request merged**